cmapPy/pandasGEXpress/tests/test_slice_gctoo.py

```diff
@@ -16,33 +16,97 @@ def setUpClass(cls):
-
-        # Outputs should be dataframes even if there is only 1 index or column
-        pd.util.testing.assert_frame_equal(out_gct.data_df, self.in_gct.data_df.iloc[[0, 2], :])
-        pd.util.testing.assert_frame_equal(out_gct.row_metadata_df, self.in_gct.row_metadata_df.iloc[[0, 2], :])
-        pd.util.testing.assert_frame_equal(out_gct.col_metadata_df, self.in_gct.col_metadata_df)
+        # Error if resulting GCT is empty
+        with self.assertRaises(AssertionError) as e:
+            sg.slice_gctoo(self.in_gct, rid=["bad"], cid=["x", "y"])
+        self.assertIn("Slicing yielded an", str(e.exception))
 
-    def test_slice_cid_and_col_bool(self):
         # cid and col_bool should not both be provided
         with self.assertRaises(AssertionError) as e:
             sg.slice_gctoo(self.in_gct, cid=["e", "f", "g"], col_bool=[True, True, False])
-        self.assertIn("cid and col_bool", str(e.exception))
+        self.assertIn("Only one of cid,", str(e.exception))
+
+        # Providing all 3 row inputs is also bad!
+        with self.assertRaises(AssertionError) as e:
+            sg.slice_gctoo(self.in_gct, rid="blah", ridx="bloop", row_bool="no!")
+        self.assertIn("Only one of rid,", str(e.exception))
+
+        # happy path
+        out_g = sg.slice_gctoo(self.in_gct, rid=["d", "a", "b"], cidx=[0],
+                               exclude_rid=["a"])
+        pd.util.testing.assert_frame_equal(out_g.data_df, self.in_gct.data_df.iloc[[1, 3], [0]])
 
+    def test_get_rows_to_keep(self):
+
+        # rid must be a list
+        with self.assertRaises(AssertionError) as e:
+            sg.get_rows_to_keep(self.in_gct, rid="bad")
+        self.assertIn("rid must be a list", str(e.exception))
+
+        # bools
+        out_rows = sg.get_rows_to_keep(self.in_gct, row_bool=[True, True, True, False])
+        self.assertItemsEqual(out_rows, ["a", "b", "c"])
+
+        # rid and exclude_rid
+        out_rows2 = sg.get_rows_to_keep(self.in_gct, rid=["a", "c", "d"], exclude_rid=["d"])
+        self.assertItemsEqual(out_rows2, ["a", "c"])
+
+        # keep all rows
+        out_rows3 = sg.get_rows_to_keep(self.in_gct)
+        self.assertItemsEqual(out_rows3, ["a", "b", "c", "d"])
+
+        with self.assertRaises(AssertionError) as e:
+            sg.get_rows_to_keep(self.in_gct, row_bool=[True, False, True])
+        self.assertIn("row_bool must have length", str(e.exception))
+
+        with self.assertRaises(AssertionError) as e:
+            sg.get_rows_to_keep(self.in_gct, ridx=[True, False, True])
+        self.assertIn("ridx must be a list of integers", str(e.exception))
+
+        with self.assertRaises(AssertionError) as e:
+            sg.get_rows_to_keep(self.in_gct, ridx=[0, 2, 5])
+        self.assertIn("ridx contains an integer", str(e.exception))
+
+    def test_get_cols_to_keep(self):
+        # N.B. annoying that we have two extremely similar but separate methods
```

Comment:
N.B. Decided not to do this, because it's not just error messages that differ. There are a few places where we have to look in the index or columns depending on the method, and it seems to me that two separate methods is better than many "if-else"s.

Reply:
Ok, seems reasonable. Thanks for looking into it!